File: 297_serialize_and_deserialize_binary_tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Codec {
    public static final String NULL_STATE = "N";
    public static final String SPLITER = ",";

    public String serialize(TreeNode root) {
        StringBuilder sb = new StringBuilder();
        buildString(root, sb);
        return sb.toString();
    }

    private void buildString(TreeNode node, StringBuilder sb) {
        if (node == null){
            sb.append(NULL_STATE).append(SPLITER);
        }else {
            sb.append(node.val).append(SPLITER);
            buildString(node.left, sb);
            buildString(node.right, sb);
        }
    }

    public TreeNode deserialize(String data) {
        String[] content = data.split(SPLITER);
        Deque<String> queue = new LinkedList<>();
        queue.addAll(Arrays.asList(content));
        TreeNode root = buildTree(queue);
        return root;
    }

    private TreeNode buildTree(Deque<String> queue) {
        String s = queue.remove();
        if (s.equals(NULL_STATE)) {
            return null;
        } else {
            TreeNode node = new TreeNode(Integer.valueOf(s));
            node.left = buildTree(queue);
            node.right = buildTree(queue);
            return node;
        }
    }
}

// Your Codec object will be instantiated and called as such:
// Codec codec = new Codec();
// codec.deserialize(codec.serialize(root));

File: python/297_serialize_and_deserialize_binary_tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Codec:

    def serialize(self, root):
        """Encodes a tree to a single string.

        :type root: TreeNode
        :rtype: str
        """
        q = collections.deque()
        q.append(root)
        res = []

        while q:
            curr = q.popleft()
            if curr:
                res.append(curr.val)
                q.append(curr.left)
                q.append(curr.right)
            else:
                res.append(None)
        return str(res)

    def deserialize(self, data):
        """
        apply bfs and index for decoding string
        """
        orig = eval(data)
        root = TreeNode(orig[0]) if orig and orig[0] != None else None
        i = 1
        q = collections.deque([root])
        while q:
            node = q.popleft()
            if node is not None:
                # ensure it has child
                node.left = TreeNode(orig[i]) if orig[i] != None else None
                node.right = TreeNode(orig[i + 1]) if orig[i + 1] != None else None
                i += 2
                q.extend([node.left, node.right])
        return root






# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.deserialize(codec.serialize(root))
